cpp/src/arrow/c/bridge.cc

```diff
@@ -663,6 +672,112 @@ Status ExportRecordBatch(const RecordBatch& batch, struct ArrowArray* out,
   return Status::OK();
 }
 
+//////////////////////////////////////////////////////////////////////////
+// C device arrays
+
+Status ValidateDeviceInfo(const ArrayData& data, DeviceType* device_type,
+                          int64_t* device_id) {
+  for (const auto& buf : data.buffers) {
+    if (!buf) {
+      continue;
+    }
+
+    if (*device_type == DeviceType::UNKNOWN) {
```

Comment:
Pretty sure this loop would fail to catch a discrepancy if all buffers had `DeviceType::UNKNOWN` except for the last one.

Reply:
But that is fine I think. The constraint imposed here is that non-UNKNOWN buffers have to be distinct. So only one different from UNKNOWN is trivially distinct.

Reply:
If that's the case, then the function would return `Invalid` if any non-UNKNOWN buffers were "seen" before an UNKNOWN buffer (unless I'm totally missing something).

Reply:
The `UNKNOWN` is coming from line 710, it's what we initialize the value with before we make the recursive call. @pitrou is your suggestion that I should remove the `UNKNOWN` value from the enum and just use `std::optional` everywhere?

Reply:
Yes, please remove `UNKNOWN` from the enum as it shouldn't be a legal value for it.